Ctl 005 create new contact screen

heupotato opened the pull request:
@lngyeen Pull Request: Screen for creating new contact (UI only)

heupotato committed:
Update navigator to new contact screen

heupotato committed:
Create new contact screen

lngyeen commented on lib/ui/screens/new_contact_screen.dart:
```diff
@@ -0,0 +1,149 @@
+    late String email; 
+    late String address; 
+
+    List<String> genderList = ["Male", "Female", "Others"]; 
```
Should create enum type like this:
```suggestion
    
enum Gender { male, female, others }

extension GenderName on Gender {
  String get name {
    switch (this) {
      case Gender.male:
        return 'Male';
      case Gender.female:
        return 'Female';
      case Gender.others:
        return 'Others';
    }
  }
}
```

lngyeen commented on lib/ui/screens/new_contact_screen.dart:
```diff
@@ -0,0 +1,149 @@
+    Widget lastNameField(){
+        return TextFormField(
+            decoration: InputDecoration(labelText: "Last name"),
+            validator: (String ? value) {
```
Should create a Validator class, put it in /lib/services/validator.dart

```suggestion
class Validator {
  static bool isEmail(String input) {
    /// Put validation here
  }

  static bool isPhoneNumber(String input) {
    /// Put validation here
  }
}

```

lngyeen commented on lib/ui/screens/new_contact_screen.dart:
```diff
@@ -0,0 +1,149 @@
+                            emailField(), 
+                            addressField(), 
+                            SizedBox(height: 100), 
+                            ElevatedButton(
```
If you use this type of Button with your own custom style in many places, create your own custom button widget.
Should named 'CustomElevatedButton' and put it in /lib/widgets/buttons/custom_elevated_button.dart.
Provide some properties in CustomElevatedButton constructor for example [title/color/fontSize/onPressed].

lngyeen commented on lib/ui/screens/new_contact_screen.dart:
```diff
@@ -0,0 +1,149 @@
+import 'package:flutter/material.dart';
```
UI Error: 
![Simulator Screen Shot - iPhone 11 Pro Max - 2021-06-23 at 17 22 51](https://user-images.githubusercontent.com/60918798/123081142-d8ccd080-d447-11eb-8bda-09884b9b74eb.png)

lngyeen commented on lib/ui/screens/new_contact_screen.dart:
```diff
@@ -0,0 +1,149 @@
+                                    "Submit", 
+                                    style: TextStyle(color: Colors.white, fontSize: 16, fontWeight: FontWeight.bold)), 
+                                onPressed: () => {
+                                        if (!formKey.currentState!.validate()){
```
Should not use force unwrap for an optional variable.
if closure should be
`if (formKey.currentState?.validate() ?? false) { ///Add new contact }`

heupotato committed:
Change List<String> Gender into enum Gender

heupotato committed:
Create class Validator to validate input fields

Only first name, last name and phone number required validating. Therefore I created validator methods for these only 3 fields

heupotato committed:
Update new_contact_screen.dart

heupotato committed:
Update workspace.xml

lngyeen commented on lib/ui/screens/new_contact_screen.dart:
```diff
@@ -0,0 +1,152 @@
+import 'package:flutter/material.dart';
+import 'package:flutter_contact_list/services/validator.dart';
+
+class NewContact extends StatefulWidget {
```
Rename class name to 'NewContactScreen'

lngyeen commented on lib/ui/screens/new_contact_screen.dart:
```diff
@@ -0,0 +1,152 @@
+    late String firstName; 
+    late String lastName; 
+    late String phoneNumber; 
+    String gender = Gender.others.name;
```
'gender' should be enum type (Gender)

lngyeen commented on lib/ui/screens/contact_list_screen.dart:
```diff
@@ -1,6 +1,7 @@
 import 'package:flutter_contact_list/ui/screens/contact_detail_screen.dart';
+import 'package:flutter_contact_list/ui/screens/new_contact_screen.dart';
 
 class ContactList extends StatefulWidget {
```
Change class name to ContactListScreen

lngyeen commented on lib/services/validator.dart:
```diff
@@ -0,0 +1,23 @@
+        ? "Last name is required" : null;
+  }
+
+  static String ? isPhoneNumber (String phone){
```
Should add phone number Regex for better validation

lngyeen commented on lib/ui/screens/contact_list_screen.dart:
```diff
@@ -64,7 +65,10 @@ class _ContactListState extends State<ContactList> {
     @override
     Widget build(BuildContext context) {
         _pushAdd(){
```
function should be, change move it to a dedicated function (outside this build function):

```suggestion
        _goToAddNewContact(BuildContext context) {
    Navigator.push(
        context, MaterialPageRoute(builder: (context) => NewContact()));
  }
```

heupotato committed:
Make changes in new contact screem

Consist of:
- Change class' name
- Change variable gender's type from String to Gender
- Remove old code for elevated button

heupotato committed:
Update main.dart

Cause class' name of contact-list-screen has changed

heupotato committed:
Update phone RegEx for better validation

heupotato committed:
Create custom_elevated_button.dart

heupotato committed:
Update contact_list_screen.dart

Consist of:
- Change class' name in to ContactListScreen
- move _pushAdd outside build Widget

heupotato committed:
Update contact_list_screen.dart

lngyeen commented on lib/services/validator.dart:
```diff
@@ -0,0 +1,22 @@
+class Validator{
+  static String? isFirstName(String firstname){
```
If return a string value, function name should not start with 'is'. Change function name to 
static String? firstName(String firstname)

Do the same for other validators.

heupotato committed:
Rename validator methods

heupotato committed:
Update on validator
